src/flow_graph.py

```diff
@@ -1013,10 +1032,169 @@ def terminate_infinite_loops(nodes: List[Node]) -> None:
         compute_relations(nodes)
 
 
+@dataclass(frozen=True)
+class InstrRef:
+    node: Node
+    index: int
+
+    def instruction(self) -> "Instruction":
+        return self.node.block.instructions[self.index]
+
+    def replace_instruction(self, instr: Instruction) -> None:
+        self.node.block.instructions[self.index] = instr
+
+    def __repr__(self) -> str:
+        line = self.instruction().meta.lineno
+        base = f"{self.node.block.index}.{self.index}"
+        return f"{base} (line {line})" if line else base
+
+
+Reference = Union[InstrRef, str]
```

Comment:
fleeting thought: I wonder if this should be folded into InstrRef. `str` kinda represents a reference to a pseudo-instruction at the very start that sets up constant registers. We could make that instruction explicit even, though I suspect it's not a good idea

Reply:
I did go back-and-forth on folding `str` into `InstrRef` or not (at which point it would just be called `Reference`). I decided to represent these "constant registers" with a bare `str` for two reasons: 
- Python has poor support for sum types. It's awkward to make every field in the dataclass `Optional[...]` and frequently assert that the fields you expect are not `None`.
- There are times when I want to make it explicit that I have an *instruction* reference, and I expect  that the `.instruction()` method should return a valid `Instruction`. (The constant registers have nothing to return there)

I also decided to use a bare `str` instead of a complex dataclass/enum because the values are pretty static and aren't analyzed further. (And as said before, it's very cumbersome to define/use sum types.)